### DISCUSSION TITLE
关于PaoPaoGateWay的Full Cone NAT类型

### DISCUSSION THREAD

**Discussion by hsj1992** [kkkgo/PaoPaoGateWay#46] [问和答] [answered]
想请问PaoPaoGateWay实现的`Full Cone NAT`效果，用[NatTypeTester](https://github.com/HMBSbige/NatTypeTester)测试经过PaoPaoGateWay的STUN服务器`stunserver.stunprotocol.org`，映射行为和过滤行为都是`EndpointIndependent`吗？

有这疑问是因为使用的主路由ROS系统，在开启了`Endpoint-Independent NAT`后，

> ```
> /ip firewall nat
> add action=endpoint-independent-nat chain=srcnat out-interface=pppoe-out1 protocol=udp
> add action=endpoint-independent-nat chain=dstnat in-interface=pppoe-out1 protocol=udp
> ```

使用软件测试，直连的STUN服务器`stun.qq.com`的两个行为都是`EndpointIndependent`，但经过PaoPaoGateWay的STUN服务器`stunserver.stunprotocol.org`测试，显示的映射行为是`AddressDependent`,过滤行为是`EndpointIndependent`，想知道是否是我ROS设置的问题。

**Comment by kkkgo** [comment]
你可以把电脑网关和DNS都设置成PaoPaoGateWay测试下Full Cone NAT类型。主路由设置不太清楚，如果成功了可以分享下。

**Reply by hsj1992** [reply]
将DNS和电脑网关都设置成PaoPaoGateWay测试NAT类型，两个行为都是EndpointIndependent，是Full Cone NAT没错。
(暂时先把需要使用Full Cone NAT的设备用DHCP option设置网关成PaoPaoGateWay了）

**Reply by hsj1992** [reply, marked as answer]
使用局域网的一台 Linux 设备进行测试（使用`stun-client`），发现了问题：
执行命令`stun stunserver.stunprotocol.org -v`，然后按常理，如果是 NAT1 的 EIM NAT 的话，显示的数个`mapped address`应该都会是同一个 proxy 的 ip 地址。但是其中的一个 mapped address 出现了 direct 的 ip 地址。
我做了个尝试，在主路由 ros 里将该 stun 服务器的国外 ip 地址网段`3.132.0.0/16`和`3.132.0.0/16`静态路由指向旁路网关，对该国外 stun 服务器再次进行测试，就好了,mapping 和 filter 都是 endpoint-independent 了。

所以应该是 fakeip DNS 分流之外的直接指向IP的问题？
就像要使用 tg 还需要给它的 CIDR 网段写静态路由指向旁路网关一样，NAT 类型探测也会有设备直接指向 IP 而没经过 fakeip DNS 的场景？
然后导致 mapped address 有不同，导致`mapped IP same = 0`，从而显示`AddressDependent Mapping`。
所以当我设备的网关地址直接设置为旁路网关时候是 EIM NAT ，因为此时设备没经过域名指向 IP 也是走的旁路网关。

**Reply by kkkgo** [reply]
按你的描述，那就是检测程序就是这么判定的，FakeIP肯定实际IP不一致，跟实际效果无关。
实际上，full cone nat的应用场景也是P2P联机，也不会用到域名，P2P添加静态路由也没意义，所以应用场景也就是把设备网关直接指向ppgw。
